packages/bundlers/experimental/src/ExperimentalBundler.js

```diff
@@ -895,39 +892,57 @@ function createIdealGraph(
 
     // Finally, filter out bundleRoots (bundles) from this assets
     // reachable if they are subgraphs, and reuse that subgraph bundle
-    // by drawing an edge. Essentially, if two bundles within an assets
+    // by drawing an edge. Essentially, if two bundles within an asset's
     // reachable array, have an ancestor-subgraph relationship, draw that edge.
     // This allows for us to reuse a bundle instead of making a shared bundle if
     // a bundle represents the exact set of assets a set of bundles would share
 
     // if a bundle b is a subgraph of another bundle f, reuse it, drawing an edge between the two
-    reachable = reachable.filter(b => {
-      if (b.env.isIsolated()) {
-        return true;
-      }
-      let toKeep = true;
-      if (bundles.has(asset.id)) {
-        toKeep = false;
-        bundleGraph.addEdge(
-          nullthrows(bundles.get(b.id)),
-          nullthrows(bundles.get(asset.id)),
-        );
+    let canReuse: Set<BundleRoot> = new Set();
+    for (let candidateSourceBundleRoot of reachable) {
+      let candidateSourceBundleId = nullthrows(
+        bundles.get(candidateSourceBundleRoot.id),
+      );
+      if (candidateSourceBundleRoot.env.isIsolated()) {
+        continue;
       }
-      for (let f of reachable) {
-        if (b === f) continue;
-        let fReachable = getReachableBundleRoots(f, reachableRoots).filter(
-          b => !ancestorAssets.get(b)?.has(f),
-        );
-        if (fReachable.indexOf(b) > -1) {
-          toKeep = false;
-          bundleGraph.addEdge(
-            nullthrows(bundles.get(b.id)),
-            nullthrows(bundles.get(f.id)),
-          );
+      let reuseableBundleId = bundles.get(asset.id);
+      if (reuseableBundleId != null) {
+        canReuse.add(candidateSourceBundleRoot);
+        bundleGraph.addEdge(candidateSourceBundleId, reuseableBundleId);
+
+        let reusableBundle = bundleGraph.getNode(reuseableBundleId);
+        invariant(reusableBundle !== 'root' && reusableBundle != null);
+        reusableBundle.sourceBundles.add(candidateSourceBundleId);
+      } else {
+        // Asset is not a bundleRoot, but if its parent bundle can be
```

Comment:
Only the direct parent? Is there any case where we'd have to do this recursively?

Reply:
So actually I think the comment isn't the best. Essentially what we're trying to do here is: 

Of the reachable assets for an asset, is there some common subset (under a bundleRoot) that may be reused. 
So I guess it's not so much as a parent, as it is any bundleRoot that synchronously imports the asset in question. (which to me is kind of a parent? Just not the **only** parent)

So I guess the better way to phrase would be "if **a bundleRoot of a bundle that this asset is in,** can be reused by some other bundleRoot in the asset's reachable, reuse it"

Reply:
And it's important to note, this code path you're talking about was only added because we can't guarantee that we process a bundleRoot before any non-bundleRoot assets that are in it, when placing assets into bundles. 
So like if we process a non-bundleRoot asset, we need to make sure we don't prematurely place it in, where it could've been reused by some other bundleRoot

Reply:
Diagrams for this section might be important 😄